velox/exec/Window.cpp

```diff
@@ -41,16 +41,78 @@ void initKeyInfo(
   }
 }
 
-void checkDefaultWindowFrame(const core::WindowNode::Function& windowFunction) {
-  VELOX_CHECK_EQ(
-      windowFunction.frame.type, core::WindowNode::WindowType::kRange);
-  VELOX_CHECK_EQ(
-      windowFunction.frame.startType,
-      core::WindowNode::BoundType::kUnboundedPreceding);
-  VELOX_CHECK_EQ(
-      windowFunction.frame.endType, core::WindowNode::BoundType::kCurrentRow);
-  VELOX_CHECK_EQ(windowFunction.frame.startValue, nullptr);
-  VELOX_CHECK_EQ(windowFunction.frame.endValue, nullptr);
+vector_size_t findFrameEndPointKPreceding(
+    core::WindowNode::WindowType type,
+    vector_size_t functionNumber,
+    vector_size_t partitionStartRow,
+    vector_size_t /*partitionEndRow*/,
+    vector_size_t currentRow,
+    std::optional<vector_size_t> offsetValue = 0) {
+  if (!offsetValue || type != core::WindowNode::WindowType::kRows) {
+    VELOX_FAIL("k preceding as frame start is allowed only in ROWS mode");
+  }
+  return std::max(
+      currentRow - vector_size_t(offsetValue.value()), partitionStartRow);
+}
+
+vector_size_t findFrameEndPointCurrentRow(
+    core::WindowNode::WindowType type,
+    vector_size_t /*functionNumber*/,
+    vector_size_t /*partitionStartRow*/,
+    vector_size_t /*partitionEndRow*/,
+    vector_size_t currentRow,
+    std::optional<vector_size_t> offsetValue = 0) {
+  return currentRow;
+}
+
+vector_size_t findFrameEndPointKFollowing(
+    core::WindowNode::WindowType type,
+    vector_size_t functionNumber,
+    vector_size_t /*partitionStartRow*/,
+    vector_size_t partitionEndRow,
+    vector_size_t currentRow,
+    std::optional<vector_size_t> offsetValue = 0) {
+  if (!offsetValue || type != core::WindowNode::WindowType::kRows) {
+    VELOX_FAIL("k preceding as frame end is allowed only in ROWS mode");
+  }
+  return std::min(
+      currentRow + vector_size_t(offsetValue.value()), partitionEndRow);
+}
+
+void setWindowFrameStart(
```

Comment:
This function can return a WindowFrameFunctionPtr instead of passing by reference.